drivers/mipi_dbi/mipi_dbi_spi.c

```diff
@@ -0,0 +1,294 @@
+			gpio_pin_set_dt(&config->cmd_data, config->cd_invert);
+			ret = spi_write(config->spi_dev, &dbi_config->config,
+					&buf_set);
+			if (ret < 0) {
+				goto out;
+			}
+		}
+
+		if (len > 0) {
+			buffer.buf = (void *)data_buf;
+			buffer.len = len;
+
+			gpio_pin_set_dt(&config->cmd_data, !config->cd_invert);
+			ret = spi_write(config->spi_dev, &dbi_config->config,
+					&buf_set);
+			if (ret < 0) {
+				goto out;
+			}
+		}
+	} else {
+		/* Otherwise, unsupported mode */
+		ret = -ENOTSUP;
+	}
+out:
+	k_spin_unlock(&data->lock, spinlock_key);
+	return ret;
+}
+
+static int mipi_dbi_spi_command_write(const struct device *dev,
+				      const struct mipi_dbi_config *dbi_config,
+				      uint8_t cmd, const uint8_t *data_buf,
+				      size_t len)
+{
+	return mipi_dbi_spi_write_helper(dev, dbi_config, true, cmd,
+					 data_buf, len);
+}
+
+static int mipi_dbi_spi_write_transfer(const struct device *dev,
+				       const struct mipi_dbi_config *dbi_config,
+				       const uint8_t *data_buf, size_t len)
+{
+	return mipi_dbi_spi_write_helper(dev, dbi_config, false, 0x0,
+					 data_buf, len);
+}
+
+static int mipi_dbi_spi_command_read(const struct device *dev,
+				     const struct mipi_dbi_config *dbi_config,
+				     uint8_t cmd, uint8_t *response,
+				     size_t len)
+{
+	const struct mipi_dbi_spi_config *config = dev->config;
+	struct mipi_dbi_spi_data *data = dev->data;
+	struct spi_buf buffer;
+	struct spi_buf_set buf_set = {
+		.buffers = &buffer,
+		.count = 1,
+	};
+	int ret;
+	k_spinlock_key_t spinlock_key = k_spin_lock(&data->lock);
+
+	if (dbi_config->mode == MIPI_DBI_MODE_SPI_3WIRE &&
+	    IS_ENABLED(CONFIG_MIPI_DBI_SPI_3WIRE)) {
+		struct spi_config cmd_config, data_config;
+		/* We have to emulate 3 wire mode by packing the data/command
+		 * bit into the upper bit of the SPI transfer.
+		 * switch SPI to 9 bit mode, and write the transfer
+		 */
+		memcpy(&cmd_config, &dbi_config->config, sizeof(cmd_config));
+		memcpy(&data_config, &dbi_config->config, sizeof(data_config));
+		cmd_config.operation &= ~SPI_WORD_SIZE_MASK;
+		cmd_config.operation |= SPI_WORD_SET(9);
+		data_config.operation &= ~SPI_WORD_SIZE_MASK;
+		data_config.operation |= SPI_WORD_SET(8);
+
+		/* Send command */
+		if (config->cd_invert) {
+			data->spi_byte = MIPI_DBI_DC_BIT | cmd;
+		} else {
+			data->spi_byte = cmd;
+		}
+		buffer.buf = &data->spi_byte;
+		buffer.len = 1;
+		ret = spi_write(config->spi_dev, &cmd_config, &buf_set);
+		if (ret < 0) {
+			goto out;
+		}
+		/* Now, we can switch to 8 bit mode, and read data */
+		buffer.buf = (void *)response;
+		buffer.len = len;
+		ret = spi_read(config->spi_dev, &data_config, &buf_set);
+	} else if (dbi_config->mode == MIPI_DBI_MODE_SPI_4WIRE) {
+		/* 4 wire mode is much simpler. We just toggle the
+		 * command/data GPIO to indicate if we are sending
+		 * a command or data
+		 */
+		buffer.buf = &cmd;
+		buffer.len = sizeof(cmd);
+
+		gpio_pin_set_dt(&config->cmd_data, config->cd_invert);
```

Comment:
How about not using `cd-invert` here and changing the documentation specifying that cd-invert is only used if the device is in 4 wire mode? Seems like you are putting on polarity inversion fight here :-)

Reply:
Looking at the MIPI DBI spec closer, you're right- compliant displays should treat a low D/CX signal/bit as a command, and high D/CX signal/bit as data.

Given that, we could probably drop this property entirely. The only potential downside here is that a user running the display in 3 wire mode would have no way to invert the D/CX bit if their display didn't comply to the MIPI DBI standard.

This is such an edge case though that I feel dropping the `cd-invert` property for the time being makes sense. We can do the same for the `rst-invert`, since the spec specifies it is active low

Reply:
Oh... I'm not sure that's what I mean but... cool! :-)